en/lessons/basics/control-structures.md

```diff
@@ -1,207 +1,246 @@
 ---
-version: 1.1.1
+version: 1.0.0
 title: Control Structures
 ---
 
-In this lesson we will look at the control structures available to us in Elixir.
+Recall that Haskell functions are *definitions*, not lists of instructions
 
-{% include toc.html %}
+In imperative programming, control structures are used to decide which block of
+code should be executed
 
-## `if` and `unless`
+In Haskell, they are used instead to decide how a certain value should be
+defined
 
-Chances are you've encountered `if/2` before, and if you've used Ruby you're familiar with `unless/2`.
-In Elixir they work much the same way but they are defined as macros, not language constructs. You can find their implementation in the [Kernel module](https://hexdocs.pm/elixir/Kernel.html).
+__Note__: Usually control structures are used directly in the definition of a 
+function, but they can also be used in the definition of any intermediate
+value inside a function (we will see examples of both)
 
-It should be noted that in Elixir, the only falsey values are `nil` and the boolean `false`.
+Let's have a look at the control structures available in Haskell, and some
+examples of how they work
 
-```elixir
-iex> if String.valid?("Hello") do
-...>   "Valid string!"
-...> else
-...>   "Invalid string."
-...> end
-"Valid string!"
+{% include toc.html %}
 
-iex> if "a string value" do
-...>   "Truthy"
-...> end
-"Truthy"
-```
+## Extensions and Imports
 
-Using `unless/2` is like `if/2` only it works on the negative:
+You will need to add the following lines to the top of your Haskell source 
+file (`.hs`) to be able to run these examples
 
-```elixir
-iex> unless is_integer("hello") do
-...>   "Not an Int"
-...> end
-"Not an Int"
+```haskell
+{-# LANGUAGE OverloadedStrings #-}
+
+import Data.Text ( Text )
```

Comment:
Wouldn't it be preferable to keep inside the Prelude namespace as much as possible? I know `String` are bad for many things, but don't they offer a simple and intuitive interface for teaching to new users?

Reply:
Reasoning about lazy linked lists when speaking of text is far from being intuitive.

Reply:
So this document is solely about syntax, and the syntax/interface provided by Strings is extremely intuitive. It's only their semantics that is not. But since the semantics of Haskell in general is not intuitive at all, why not handle semantic difficulties separately?

Reply:
Because then you have to undo the intuition that you promote afterwards. We already have enough space leak issues with people blindingly using `String` because they are "so intuitive to use", I don't want to contribute legitimising `String` as a valid textual datatype: It is not and we should stop fooling ourselves.

Reply:
I noticed that the very first lesson uses `Text`, and I figured that was intentional. So I did the same here. Also using the `<>` operator is just as intuitive to me as using `++` with `String`

Reply:
> I noticed that the very first lesson uses `Text`, and I figured that was intentional. So I did the same here. Also using the `<>` operator is just as intuitive to me as using `++` with `String`

I very much understand the idea of promoting the `text` package. I really do. Everyone doing anything serious in Haskell uses it instead of `String` for almost anything regarding text contents. Also `Data.Text` might become part of the Prelude one day (after we move from Utf16 to Utf8 probably). But if you look at any lauded pedagogical material for Haskell (_Haskell Programming From First Principles_, more controversially _Get Programming With Haskell_ for example), the `Text` type is introduced _much later_ and for reasons. (quoting from Kurt Manning:

> While learning Haskell, String is useful for two reasons. First, as mentioned, many of the basic string utilities are baked into the standard Prelude. Second, lists are to Haskell what arrays are to C. Many concepts in Haskell are nicely demonstrated with lists, and strings are useful lists. For learning purposes, feel free to stick with String.

To me the first reason means: now you need to `import Data.Text` everywhere, which clutters the material, why the second reason means that you can easily use String to illustrate lists -- including the list monad. But there are other reasons; you need to explain why one should import the strict and not the lazy version, or vice-versa; or why `Data.Text` has its own separate functions different from the Prelude functions; or why `Data.Text` instead of `ut8-bytestring` or `Data.Vector.Unboxed.Vector Char`.

In other word, it seems to me that using `Text` everywhere is a bit of a premature optimization driven by ideology rather than pragmatism. A better approach in my opinion would be to:

1) use String whenever the focus is on cons-able datastructures
2) use Text whenever the focus is on efficient text processing.

Reply:
Well, you're entitled to your opinion. :)

Reply:
> Well, you're entitled to your opinion. :)

Can we get the perspective of the folks supporting the Haskell School project on this?

Reply:
Haskell School does need to keep its own style (as a reference material to get people quickly acquainted with Haskell). 

I don't feel like it's particularly productive to debate over the usage of `Text`: it's a project wide decision and debating over it slows down the progress of individual pull requests. It would be better to get to an initial release and then see whether the benefits of using `Text` have been realised. 

However, these internal decisions perhaps aren't the clearest from the outside and your review has been very helpful in showing potential misunderstandings a reader may have.

Reply:
Sure, I don't want to slow down this PR with a project-wide decision. Shall I open an Issue then?

Reply:
I think we're considering putting these decisions in the style guide, and then maybe there can be a discussion about the style guide somewhere.

Reply:
@why-not-try-calmer I share the concern of using `text` from the get-go, [as I mentioned in an earlier PR](https://github.com/haskellfoundation/HaskellSchool/pull/7#discussion_r642747500).

I encourage Haskell School government to seek a wider input on this matter, e.g. via Discourse / Reddit / mailing lists.

Reply:
I support teaching `Text` from the beginning. I learned Haskell in the last year, and it feels bad to learn topics as a beginner, only to find out that they're actually bad practice. I think there are several examples of this

1. `String` - you find out that the tools for it are clunky and that you should use `Text` instead
2. Lists - You find out that usually you should use some `Map` type instead, and that lists involve unsafe partial functions you're not supposed to use
3. Explicit recursion - You find out it can cause memory leaks, and that you rarely actually use it in practice

There are probably others. TBH these issues are why I want to contribute to this project.

Reply:
I am going to speak as the Editor-in-chief of this content platform: 

Teaching `Text` from the beginning has several advantages:
* It helps reducing cognitive load, as the learner doesn't have to "un-learn" String
* It gives the learner proper habits from the beginning
* It reduces the amount of confusion arising from the fact that `String` is simply an alias for `[Char]` and GHC doesn't always expand or retain aliases.
* The `String` utilities baked into the Prelude are simply insufficient when going beyond very simple or contrived examples, and they possess equivalents in `Text`.

As a general statement, `String` is a historical accident that we kept and I do not wish to promote something from which languages older than Haskell like Erlang have moved away.

@why-not-try-calmer @ulysses4ever You may not agree with this decision but such is life. @TrueBoxGuy is editor for the English language part of Haskell school, and I (and transitively, the executive team of the Haskell Foundation) have put trust in his judgement.

Reply:
FWIW I am strongly in favour of using `Text` over `String`.

Perhaps it is our fault for not being clear enough in some coding style document, but one of the primary motivating factors of this educational tool is to get a new haskeller up to speed and working with haskell in real world contexts quickly. To this end we did not want to delay learning to use `Text` over `String`. We also don't want to delay introducing them to the critical haskell libraries that they will inevitably need to learn. Also, if the argument is that `String` is convenient for teaching pattern matching, surely any other example of a list that is not an overloaded concept would be an equally good example?

Haskell School seeks to learn from mistakes made by other educational haskell materials (like the ones you mention @why-not-try-calmer). If you don't consider them to be mistakes, then perhaps I can appeal to the virtues of having diversity in available resources.

Finally, I would like to make an appeal to authority. @Kleidukos was designated the shepherd of this project, and TBG / myself have put in hours of time, early on, contributing to the project. All three of us came to the conclusion that using `Text` throughout is the approach we think is best.

I don't think this project should be an autocracy, if @ulysses4ever (who has also been around this project and contributing since the beginning) and `why-not-try-calmer` want to make a case for teaching `String` over `Text` I encourage you to bring it up in the Haskell Foundation slack and we can have it out there. But I hope its clear that this was not a one off decision, it is in service of our larger vision for the project. To summarize I think we should stick with the current approach since:

1. We have made the decision to prioritize getting the learner doing real world things as quickly as possible. This means introducing them to libraries outside of prelude earlier than other pedagogical materials
2. We think that, at best, we are improving on the canon of haskell educational material. At worst, we are providing diversity, which I argue is an end in its own right.
3. This decision was arrived at by people who were charged with executing the project, and who have invested their time in bringing it about. Know that we have discussed this in the past, and our decision was made based on what we understand the mandate of the project to be.

Reply:
Okay, it seems @Kleidukos and I posted at almost exactly the same time. I hope this doesn't come across as needless dog piling. Just wanted to clarify that.

Reply:
Representing strings as List-of-Char, especially in a unicode world, is one of the most common initially enticing decisions for all functional programming language implementors that in the long run turn out to be where sanity and performance go to die. 

I think Erlang mostly had it because of starting off being implemented in prolog, which really mostly didn't care about string handling as originally designed.

Any "let's implement a lisp" tutorial that's written by somebody who's worked with non-toy systems will almost certainly contain a paragraph to the effect of "we know this looks like a really tempting idea but trust us, don't".

It's unfortunate that String already stole a bunch of operator names that one might've wished one could use for Text, but that's the nature of long lived systems, and teaching newbies Text first will be a slightly steeper learning curve up front in return for much *much* less pain for them later, so honestly starting them off with String in 2021 would just be unnecessary cruelty.

Yours, mst, speaking from bitter experience over multiple FP implementations.

Reply:
Fair enough, this PR is probably not the best place for debating the presupposition -- which by now you'll know I don't accept -- that illustrating basic Haskell with `String`s entails a general endorsement or promotion of `String`s. Feel free to ignore the initial comment that spawned this discussion.